akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -380,39 +807,74 @@ private[akka] class Shard(
   private def receiveTerminated(ref: ActorRef): Unit = {
     if (handOffStopper.contains(ref))
       context.stop(self)
-    else if (idByRef.contains(ref) && handOffStopper.isEmpty)
-      entityTerminated(ref)
+    else {
+      // workaround for watchWith not working with stash #29101
```

Comment:
fixed ?

Reply:
WatchWith is fixed but we never went back to the impl that uses a custom watch message after that was merged into this branch. Perhaps that is more clear, I'll reintroduce.